README
This is the beginning of something big.


Changelog
* Note *
Pull requests will be accepted *only* against the dev branch, except for feature branch
fixes/enhancements, in which case they should be against that branch.
Please make sure any PR's comply with our coding standards policy:
http://li3.me/docs/specs/accepted/LSR-0-coding.md

New updates are grouped by person (committer/merger), at the top of the file.
All entries are dated according to merging into the code in dev branch. THIS MEANS THE CODE MAY
NOT BE PRESENT IN THE STABLE BRANCH
For details of when a commit makes it into the stable branch see the ReleaseNotes file.


2020-06-01 niel
	Mrg: dev => Latest-testing to prepare for updating stable 0.x to v0.8.22.0


2020-05-28 niel
	Chg: Update dependencies. In particular a fix for security bug in PHPMailer.

2020-01-06 niel
	Mrg: dev => Latest-testing to prepare for updating stable 0.x to v0.8.21.0


2019-10-31 niel
	Fix: Update url for IMDb images.


2019-10-24 niel
	Mrg: PR 2646, from 	JReming85 - Update Charisma theme's profileedit.tpl.
	Fix: Code Style.
2019-10-18 niel
	Fix: Another attempt to deal with the bad URL for cover images issue (2637). This time I'm not
		trying to juggle the output from the db, instead we try to make sure that new images have
		the correct IMDb id size. I will be writing a script to rename image files with id number
		less than 8 characters.
2019-10-15 niel
	Fix: Further fix the regex to be more specific.
2019-10-14 niel
	Fix: Tighten up regex used to Categorise MP3s, to avoid false positives. (Closes Issue 2642).
2019-10-08 niel
	Fix: Refactor function to check a file path for readabilty. (Closes Issue 2641)
2019-10-07 niel
	Mrg: dev => Latest-testing to prepare for updating stable 0.x to v0.8.20.0


2019-10-02 niel
	Fix: Install failing to create PROCEDURES during database initialisation.
	Chg: Update dependencies.
2019-09-19 niel
	Fix: Force Movies::getMovieRange() to return imdbid without leading zeroes. (Closes Issue 2637).
2019-09-06 niel
	Chg: Updated Lithium framework to official 1.2 version (we've been using a custom version to
		provide PHP 7.3 support).
2019-09-06 niel
	Mrg: dev => Latest-testing to prepare for updating stable 0.x to v0.8.19.0


2019-07-21 niel
	Fix: .../nzedb/NameFixer.php, Convert 'continue' to 'break' in switch statement.
2019-07-19 niel
	Mrg: dev => Latest-testing to prepare for updating stable 0.x to v0.8.18.3


2019-07-19 niel
	Fix: Missing ',' in SQL, hopefully reduce overhead formatting 'Message-Id'. Closes Issue 2633.
2019-07-18 niel
	Mrg: dev => Latest-testing to prepare for updating stable 0.x to v0.8.18.2

2019-07-18 niel
	Fix: Force NNTP headers to title case (Upper-case first letter, the rest lower-case. (Closes
		Issue 2632).
2019-07-18 niel
	Mrg: dev => Latest-testing to prepare for updating stable 0.x to v0.8.18.1


2019-07-18 niel
	Fix: Remove include_once that was made superfluous.
	Chg: 'Optimisations' for NNTP::getXOVER().
2019-07-18 niel
	Mrg: dev => Latest-testing to prepare for updating stable 0.x to v0.8.18.0


2019-07-18
	Fix: Ensure DB.php has Settings.php loaded if available.
2019-07-17 DariusIII
	Chg: Add newznab:apilimit to API and RSS XML response with limits and current api hits and grabs
2019-06-19 niel
	Mrg: PR 2631, from Wally73 - Update movie-add.php for larger IMDb IDs.
2019-06-03 niel
	Fix: Increase size of imdbid fields in movieinfo, releases, and user_movies tables to bring us
		inline with IMDb's latest entries.
2019-06-03 niel
	Mrg: dev => Latest-testing to prepare for updating stable 0.x to v0.8.17.0


2019-06-02 niel
	Add: Phinx package to handle Db updates in future.


2019-05-09 niel
	Add: Backport change from CakePHP/Laravel that copies TSV files to /tmp before trying to LOAD
		INFILE them. This should resolve any permission issues as /tmp is readable by anyone by
		default.


2019-03-25 niel
	Fix: Reduce the size of tmux table's 'value' column from 19000 to a more reasonable size (1000).
2019-03-04 niel
	Fix: Update some regexes that were using invalid syntax. This will prevent them from being
		flagged as the errors they are with PCRE2 in PHP 7.3+.
2019-03-03 niel
	Mrg: dev => Latest-testing to prepare for updating stable 0.x to v0.8.16.0


2019-03-01 niel
	Fix: Use correct object to check for video duration. Closes issue 2625
2019-02-06 niel
	Fix: Use absolute paths for version specific configuration files, in tmux.conf. Closes issue 2623.
2019-02-04 niel
	Mrg: dev => Latest-testing to prepare for updating stable 0.x to v0.8.15.0


2019-01-12 niel
	Fix: Update search suffix for poppcorn searches. Closes issue 2620.
2019-01-07 niel
	Mrg: dev => Latest-testing to prepare for updating stable 0.x to v0.8.14.0


2019-01-05 Bart39
	Fix: Invalid object type in ProcessAdditional::getMediaInfo.
2018-12-27 niel
	Fix: Use empty() to check for '' and NULL for filename in NameFixer::matchPredbFiles().
2018-12-24 niel
	Fix: Removed extra tab in categories tsv file that caused XXX->UHD to have incorrect parent ID.
	Chg: Increased size of xref field in collections and multigroup_collections to avoid truncation.
		**NOTE** There is no patch for existing installs yet. This will be coming when I get my
		backup server fixed as it handles validating the SQL etc.
2018-12-20 niel
	Fix Correct pd.predate to pd.created in viewnzb.tpl file for Alpha and Default themes
2018-12-19 niel
	Chg: Add 'z' to regexes detecting RAR parts i.e. z01 as a suffix.
2018-12-11 DariusIII
	Fix: Fix inserting of new mediainfo xml into releaseextrafull table (fixes issue #2609)
2018-12-10 niel
	Fix: PHP Fatal error: Uncaught TypeError: Argument 1 passed to nzedb\NameFixer::nzedb.
2018-12-03 niel
	Mrg: dev => Latest-testing to prepare for updating stable 0.x to v0.8.13.0


2018-11-27 niel
	Fix: New groups not respecting number of days/posts to initialise with and then starting from
		first header on the server.
2018-11-21 niel
	Fix: Correct categories table data for Book entries.
2018-11-16 niel

nzedb/TV.php
<?php
namespace nzedb;

use nzedb\db\Settings;

/**
 * Class TV
 */
class TV
{
	const SOURCE_TVDB =   1; // Scrape source was TVDB
	const SOURCE_TRAKT =  2; // Scrape source was TraktTV
	const SOURCE_TVRAGE = 3; // Scrape source was TvRage
	const SOURCE_TVMAZE = 4; // Scrape source was TVMAZE
	const SOURCE_IMDB =   5; // Scrape source was IMDB
	const SOURCE_TMDB =   6; // Scrape source was TMDB

	/**
	 * @var \nzedb\db\Settings
	 */
	public $pdo;

	/**
	 * @var bool
	 */
	public $echooutput;

	/**
	 * @param array $options Class instances / Echo to CLI.
	 */
	public function __construct(array $options = [])
	{
		$defaults = [
			'Echo'     => false,
			'Settings' => null,
		];
		$options += $defaults;
		$this->pdo = ($options['Settings'] instanceof Settings ? $options['Settings'] : new Settings());
		$this->echooutput = ($options['Echo'] && nZEDb_ECHOCLI);
		$this->catWhere = 'categoryid BETWEEN 5000 AND 5999';
	}

	public function cleanName($str)
	{
		$str = str_replace(['.', '_'], ' ', $str);

		$str = str_replace(['à', 'á', 'â', 'ã', 'ä', 'æ', 'À', 'Á', 'Â', 'Ã', 'Ä'], 'a', $str);
		$str = str_replace(['ç', 'Ç'], 'c', $str);
		$str = str_replace(['Σ', 'è', 'é', 'ê', 'ë', 'È', 'É', 'Ê', 'Ë'], 'e', $str);
		$str = str_replace(['ì', 'í', 'î', 'ï', 'Ì', 'Í', 'Î', 'Ï'], 'i', $str);
		$str = str_replace(['ò', 'ó', 'ô', 'õ', 'ö', 'Ò', 'Ó', 'Ô', 'Õ', 'Ö'], 'o', $str);
		$str = str_replace(['ù', 'ú', 'û', 'ü', 'ū', 'Ú', 'Û', 'Ü', 'Ū'], 'u', $str);
		$str = str_replace('ß', 'ss', $str);

		$str = str_replace('&', 'and', $str);
		$str = preg_replace('/^(history|discovery) channel/i', '', $str);
		$str = str_replace(['\'', ':', '!', '"', '#', '*', '’', ',', '(', ')', '?'], '', $str);
		$str = str_replace('$', 's', $str);
		$str = preg_replace('/\s{2,}/', ' ', $str);

		$str = trim($str, '\"');
		return trim($str);
	}

	public function parseNameEpSeason($relname)
	{
		$showInfo = ['name' => '', 'season' => '', 'episode' => '', 'seriesfull' => '', 'airdate' => '', 'country' => '', 'year' => '', 'cleanname' => ''];
		$matches = '';

		$following = 	'[^a-z0-9](\d\d-\d\d|\d{1,2}x\d{2,3}|(19|20)\d\d|(480|720|1080)[ip]|AAC2?|BDRip|BluRay|D0?\d' .
				'|DD5|DiVX|DLMux|DTS|DVD(Rip)?|E\d{2,3}|[HX][-_. ]?264|ITA(-ENG)?|[HPS]DTV|PROPER|REPACK|' .
				'S\d+[^a-z0-9]?(E\d+)?|WEB[-_. ]?(DL|Rip)|XViD)[^a-z0-9]';

		// For names that don't start with the title.
		if (preg_match('/[^a-z0-9]{2,}(?P<name>[\w .-]*?)' . $following . '/i', $relname, $matches)) {
			$showInfo['name'] = $matches[1];
		} else if (preg_match('/^(?P<name>[a-z0-9][\w .-]*?)' . $following . '/i', $relname, $matches)) {
		// For names that start with the title.
			$showInfo['name'] = $matches[1];
		}

		if (!empty($showInfo['name'])) {
			// S01E01-E02 and S01E01-02
			if (preg_match('/^(.*?)[^a-z0-9]s(\d{1,2})[^a-z0-9]?e(\d{1,3})(?:[e-])(\d{1,3})[^a-z0-9]/i', $relname, $matches)) {
				$showInfo['season'] = intval($matches[2]);
				$showInfo['episode'] = [intval($matches[3]), intval($matches[4])];
			}
			//S01E0102 - lame no delimit numbering, regex would collide if there was ever 1000 ep season.
			else if (preg_match('/^(.*?)[^a-z0-9]s(\d{2})[^a-z0-9]?e(\d{2})(\d{2})[^a-z0-9]/i', $relname, $matches)) {
				$showInfo['season'] = intval($matches[2]);
				$showInfo['episode'] = [intval($matches[3]), intval($matches[4])];
			}
			// S01E01 and S01.E01
			else if (preg_match('/^(.*?)[^a-z0-9]s(\d{1,2})[^a-z0-9]?e(\d{1,3})[^a-z0-9]/i', $relname, $matches)) {
				$showInfo['season'] = intval($matches[2]);
				$showInfo['episode'] = intval($matches[3]);
			}
			// S01
			else if (preg_match('/^(.*?)[^a-z0-9]s(\d{1,2})[^a-z0-9]/i', $relname, $matches)) {
				$showInfo['season'] = intval($matches[2]);
				$showInfo['episode'] = 'all';
			}
			// S01D1 and S1D1
			else if (preg_match('/^(.*?)[^a-z0-9]s(\d{1,2})[^a-z0-9]?d\d{1}[^a-z0-9]/i', $relname, $matches)) {
				$showInfo['season'] = intval($matches[2]);
				$showInfo['episode'] = 'all';
			}
			// 1x01
			else if (preg_match('/^(.*?)[^a-z0-9](\d{1,2})x(\d{1,3})[^a-z0-9]/i', $relname, $matches)) {
				$showInfo['season'] = intval($matches[2]);
				$showInfo['episode'] = intval($matches[3]);
			}
			// 2009.01.01 and 2009-01-01
			else if (preg_match('/^(.*?)[^a-z0-9](19|20)(\d{2})[^a-z0-9](\d{2})[^a-z0-9](\d{2})[^a-z0-9]/i', $relname, $matches)) {
				$showInfo['season'] = $matches[2] . $matches[3];
				$showInfo['episode'] = $matches[4] . '/' . $matches[5];
				$showInfo['airdate'] = $matches[2] . $matches[3] . '-' . $matches[4] . '-' . $matches[5]; //yy-m-d
			}
			// 01.01.2009
			else if (preg_match('/^(.*?)[^a-z0-9](\d{2})[^a-z0-9](\d{2})[^a-z0-9](19|20)(\d{2})[^a-z0-9]/i', $relname, $matches)) {
				$showInfo['season'] = $matches[4] . $matches[5];
				$showInfo['episode'] = $matches[2] . '/' . $matches[3];
				$showInfo['airdate'] = $matches[4] . $matches[5] . '-' . $matches[2] . '-' . $matches[3]; //yy-m-d
			}
			// 01.01.09
			else if (preg_match('/^(.*?)[^a-z0-9](\d{2})[^a-z0-9](\d{2})[^a-z0-9](\d{2})[^a-z0-9]/i', $relname, $matches)) {
				$showInfo['season'] = ($matches[4] <= 99 && $matches[4] > 15) ? '19' . $matches[4] : '20' . $matches[4];
				$showInfo['episode'] = $matches[2] . '/' . $matches[3];
				$showInfo['airdate'] = $showInfo['season'] . '-' . $matches[2] . '-' . $matches[3]; //yy-m-d
			}
			// 2009.E01
			else if (preg_match('/^(.*?)[^a-z0-9]20(\d{2})[^a-z0-9](\d{1,3})[^a-z0-9]/i', $relname, $matches)) {
				$showInfo['season'] = '20' . $matches[2];
				$showInfo['episode'] = intval($matches[3]);
			}
			// 2009.Part1
			else if (preg_match('/^(.*?)[^a-z0-9](19|20)(\d{2})[^a-z0-9]Part(\d{1,2})[^a-z0-9]/i', $relname, $matches)) {
				$showInfo['season'] = $matches[2] . $matches[3];
				$showInfo['episode'] = intval($matches[4]);
			}
			// Part1/Pt1
			else if (preg_match('/^(.*?)[^a-z0-9](?:Part|Pt)[^a-z0-9](\d{1,2})[^a-z0-9]/i', $relname, $matches)) {
				$showInfo['season'] = 1;
				$showInfo['episode'] = intval($matches[2]);
			}
			//The.Pacific.Pt.VI.HDTV.XviD-XII / Part.IV
			else if (preg_match('/^(.*?)[^a-z0-9](?:Part|Pt)[^a-z0-9]([ivx]+)/i', $relname, $matches)) {
				$showInfo['season'] = 1;
				$epLow = strtolower($matches[2]);
				switch ($epLow) {
					case 'i': $e = 1;
						break;
					case 'ii': $e = 2;
						break;
					case 'iii': $e = 3;
						break;
					case 'iv': $e = 4;
						break;
					case 'v': $e = 5;
						break;
					case 'vi': $e = 6;
						break;
					case 'vii': $e = 7;
						break;
					case 'viii': $e = 8;
						break;
					case 'ix': $e = 9;
						break;
					case 'x': $e = 10;
						break;
					case 'xi': $e = 11;
						break;
					case 'xii': $e = 12;
						break;
					case 'xiii': $e = 13;
						break;
					case 'xiv': $e = 14;
						break;
					case 'xv': $e = 15;
						break;
					case 'xvi': $e = 16;
						break;
					case 'xvii': $e = 17;
						break;
					case 'xviii': $e = 18;
						break;
					case 'xix': $e = 19;
						break;
					case 'xx': $e = 20;
						break;
					default:
						$e = 0;
				}
				$showInfo['episode'] = $e;
			}
			// Band.Of.Brothers.EP06.Bastogne.DVDRiP.XviD-DEiTY
			else if (preg_match('/^(.*?)[^a-z0-9]EP?[^a-z0-9]?(\d{1,3})/i', $relname, $matches)) {
				$showInfo['season'] = 1;
				$showInfo['episode'] = intval($matches[2]);
			}
			// Season.1
			else if (preg_match('/^(.*?)[^a-z0-9]Seasons?[^a-z0-9]?(\d{1,2})/i', $relname, $matches)) {
				$showInfo['season'] = intval($matches[2]);
				$showInfo['episode'] = 'all';
			}

			$countryMatch = $yearMatch = '';
			// Country or origin matching.
			if (preg_match('/\W(US|UK|AU|NZ|CA|NL|Canada|Australia|America|United[^a-z0-9]States|United[^a-z0-9]Kingdom)\W/', $showInfo['name'], $countryMatch)) {
				$currentCountry = strtolower($countryMatch[1]);
				if ($currentCountry == 'canada') {
					$showInfo['country'] = 'CA';
				} else if ($currentCountry == 'australia') {
					$showInfo['country'] = 'AU';
				} else if ($currentCountry == 'america' || $currentCountry == 'united states') {
					$showInfo['country'] = 'US';
				} else if ($currentCountry == 'united kingdom') {
					$showInfo['country'] = 'UK';
				} else {
					$showInfo['country'] = strtoupper($countryMatch[1]);
				}
			}

			// Clean show name.
			$showInfo['cleanname'] = preg_replace('/ - \d{1,}$/i', '', $this->cleanName($showInfo['name']));

			// Check for dates instead of seasons.
			if (strlen($showInfo['season']) == 4) {
				$showInfo['seriesfull'] = $showInfo['season'] . "/" . $showInfo['episode'];
			} else {
				// Get year if present (not for releases with dates as seasons).
				if (preg_match('/[^a-z0-9](19|20)(\d{2})/i', $relname, $yearMatch)) {
					$showInfo['year'] = $yearMatch[1] . $yearMatch[2];
				}

				$showInfo['season'] = sprintf('S%02d', $showInfo['season']);
				// Check for multi episode release.
				if (is_array($showInfo['episode'])) {
					$tmpArr = [];
					foreach ($showInfo['episode'] as $ep) {
						$tmpArr[] = sprintf('E%02d', $ep);
					}
					$showInfo['episode'] = implode('', $tmpArr);
				} else {
					$showInfo['episode'] = sprintf('E%02d', $showInfo['episode']);
				}

				$showInfo['seriesfull'] = $showInfo['season'] . $showInfo['episode'];
			}
			$showInfo['airdate'] = (!empty($showInfo['airdate'])) ? $showInfo['airdate'] . ' 00:00:00' : '';
			return $showInfo;
		}
		return false;
	}

	public function checkMatch($ourName, $tvrName)
	{
		// Clean up name ($ourName is already clean).
		$tvrName = $this->cleanName($tvrName);
		$tvrName = preg_replace('/ of /i', '', $tvrName);
		$ourName = preg_replace('/ of /i', '', $ourName);

		// Create our arrays.
		$ourArr = explode(' ', $ourName);
		$tvrArr = explode(' ', $tvrName);

		// Set our match counts.
		$numMatches = 0;
		$totalMatches = sizeof($ourArr) + sizeof($tvrArr);

		// Loop through each array matching again the opposite value, if they match increment!
		foreach ($ourArr as $oname) {
			if (preg_match('/ ' . preg_quote($oname, '/') . ' /i', ' ' . $tvrName . ' ')) {
				$numMatches++;
			}
		}
		foreach ($tvrArr as $tname) {
			if (preg_match('/ ' . preg_quote($tname, '/') . ' /i', ' ' . $ourName . ' ')) {
				$numMatches++;
			}
		}

		// Check what we're left with.
		if ($numMatches <= 0) {
			return false;
		} else {
			$matchpct = ($numMatches / $totalMatches) * 100;
		}

		if ($matchpct >= TvRage::MATCH_PROBABILITY) {
			return $matchpct;
		} else {
			return false;
		}
	}

	public function getGenres()
	{
		return [
				'Action', 'Adult/Porn', 'Adventure', 'Anthology',
				'Arts & Crafts', 'Automobiles', 'Buy, Sell & Trade',
				'Celebrities', 'Children', 'Cinema/Theatre', 'Comedy',
				'Cooking/Food', 'Crime', 'Current Events', 'Dance',
				'Debate', 'Design/Decorating', 'Discovery/Science', 'Drama',
				'Educational', 'Family', 'Fantasy', 'Fashion/Make-up',
				'Financial/Business', 'Fitness', 'Garden/Landscape', 'History',
				'Horror/Supernatural', 'Housing/Building', 'How To/Do It Yourself', 'Interview',
				'Lifestyle', 'Literature', 'Medical', 'Military/War',
				'Music', 'Mystery', 'Pets/Animals', 'Politics',
				'Puppets', 'Religion', 'Romance/Dating', 'Sci-Fi',
				'Sketch/Improv', 'Soaps', 'Sports', 'Super Heroes',
				'Talent', 'Tech/Gaming', 'Teens', 'Thriller',
				'Travel', 'Western', 'Wildlife'
			];
	}
}
